### PR TITLE
Use more Zephyr native headers

### PR DESCRIPTION
**Pull request by lgirdwood**
header: rtos: use rtos specific version of timer.h
    
Code can now include <rtos/timer.h> and uses native Zephyr timer API
for most uses. Some wrapping still exists which can be removed over
time.
    
Signed-off-by: Liam Girdwood <liam.r.girdwood@linux.intel.com>

### CODE REVIEW

File: zephyr/include/rtos/interrupt.h [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+static inline uint32_t interrupt_disable(uint32_t irq, void *arg)
+{
+	irq_disable(irq);
+
+	return 0;
+}
+
+/* TODO: duplicate ? disables all IRQ sources on current core */
+#define arch_interrupt_disable_mask(x)	\
+	irq_lock()
```
Comment (lyakh) [inline zephyr/include/rtos/interrupt.h:85]: this should just disable interrupts from the mask, not all? And not at the DSP but at the interrupt controller?

File: zephyr/include/rtos/interrupt.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+void interrupt_mask(uint32_t irq, unsigned int cpu);
+void interrupt_unmask(uint32_t irq, unsigned int cpu);
+void interrupt_clear_mask(uint32_t irq, uint32_t mask);
+
+/* handled by Zephyr */
+static inline void platform_interrupt_init(void) {}
+
+/* disables all IRQ sources on current core */
+#define irq_local_disable(flags) \
+	(flags = irq_lock())
```
Comment (lyakh) [inline zephyr/include/rtos/interrupt.h:97]: let's do this as
```
#define irq_local_disable(flags) do { flags = irq_lock(); } while (0)
```

File: zephyr/include/rtos/interrupt.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <stdint.h>
+
+/* to be removed once driver upstream in Zephyr */
+#define __SOF_DRIVERS_INTERRUPT_H__
+#include <platform/drivers/interrupt.h>
+
+static inline int interrupt_register(uint32_t irq, void(*handler)(void *arg), void *arg)
```
Comment (lyakh) [inline zephyr/include/rtos/interrupt.h:24]: isn't it the same code as in wrapper.c for IMX? Why do we need an additional copy?

File: zephyr/include/rtos/interrupt.h (hunk 2 above)
Comment (lgirdwood) [reply zephyr/include/rtos/interrupt.h:97]: keeping like for like atm.

File: zephyr/include/rtos/interrupt.h (hunk 3 above)
Comment (lgirdwood) [reply zephyr/include/rtos/interrupt.h:24]: The wrapper will go away over time, it may also split into smaller per feature files.

Comment (lyakh) [reply zephyr/include/rtos/interrupt.h:24]: but why is your PR keeping it in wrapper.c for IMX? It seems the same, so they can use the version in the header too?

File: zephyr/include/rtos/interrupt.h (hunk 1 above)
Comment (lgirdwood) [reply zephyr/include/rtos/interrupt.h]: actually done this the other way around now and define irq_lock() for SOF users (and its used in 1 place with 0xffffffff).

File: zephyr/include/rtos/interrupt.h (hunk 2 above)
Comment (andyross) [reply zephyr/include/rtos/interrupt.h:93]: Actually this wants to be "`arch_irq_lock()`".  On a single processor system, they're the same.  But Zephyr historically used this for global locking within drivers, which on SMP doesn't work (the fact that your CPU won't be interrupted doesn't mean no one's will).   So when CONFIG_MP_NUM_CPUS > 1, the IRQ lock becomes a somewhat complicated global spinlock, which almost certainly isn't what's desired by a API named `irq_local_disable()`.

The underlying IRQ masking step is provided by the architecture layer and still available as `arch_irq_lock()`, where it's used in a few places by the kernel (spinlocks being the big one).

Comment (lgirdwood) [reply zephyr/include/rtos/interrupt.h]: Thanks @andyross changes made to use ```arch_irq_lock()```

File: xtos/include/rtos/interrupt.h [hunk 4]
```diff
@@ -135,6 +135,11 @@ void interrupt_unregister(uint32_t irq, const void *arg);
 uint32_t interrupt_enable(uint32_t irq, void *arg);
 uint32_t interrupt_disable(uint32_t irq, void *arg);
 
+/* Zephyr compat */
+#if !defined(__ZEPHYR__)
```
Comment (lyakh) [inline xtos/include/rtos/interrupt.h:139]: this header is under xtos/include now, can it still be included for Zephyr?

**Review by lyakh — APPROVED** [review]
I think `#if !defined(__ZEPHYR__)` in xtos/include/.../interrupt.h isn't needed, would be good to remote it before merging or as an incremental commit

File: src/ipc/ipc3/handler.c [hunk 5]
```diff
@@ -661,7 +661,7 @@ static int ipc_pm_context_save(uint32_t header)
 	/* TODO: mask ALL platform interrupts except DMA */
 
 	/* mask all DSP interrupts */
-	arch_interrupt_disable_mask(0xffffffff);
+	arch_irq_lock();
```
Comment (kv2019i) [inline src/ipc/ipc3/handler.c:664]: This looks a bit weird as unmatched lock pair, but I guess this is ok given we are going to D3. OS size code to put system to D3 might probably be a more appropriate place to mask the interrupts (versus the IPC app).

File: src/audio/buffer.c [hunk 6]
```diff
@@ -7,7 +7,7 @@
 
 #include <sof/audio/buffer.h>
 #include <sof/audio/component.h>
-#include <sof/drivers/interrupt.h>
+#include <rtos/interrupt.h>
```
Comment (kv2019i) [inline src/audio/buffer.c:10]: This actually seems a bit wrong. buffer.c (and dai.c, copier.c) should have no need to use the interrupt interface directly. Not necessarily for this PR (rtos/interrupt.h is actually better), but we could probably prune the dependencies from audio code. It should really be the XTOS drivers that need these interfaces.

File: zephyr/include/rtos/alloc.h [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+#include <rtos/bit.h>
+#include <rtos/string.h>
+#include <sof/trace/trace.h>
+#include <user/trace.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+/** \addtogroup alloc_api Memory Allocation API
```
Comment (kv2019i) [inline zephyr/include/rtos/alloc.h:17]: This doesn't seem right. rmalloc() and friends is really a SOF specific "audio customer allocator" built on top of the RTOS allocator. I don't think this should be under rtos moniker, but rather a SOF service. 

We can discuss whether app code should use Zephyr allocator directly, but currently that is not even possible as you need some intermedia to set up the heaps in the way SOF expects to (currently done in wrapper.c).

File: xtos/include/rtos/interrupt.h (hunk 4 above)
Comment (lgirdwood) [reply xtos/include/rtos/interrupt.h:139]: yes for the moment - unless Zephyr native headers Kconfig is selected (but this forces out needed headers atm).

**Review by kv2019i — COMMENTED** [review]
Ok too late, this was merged while I was reviewing. I'm not 100% sure of the changes. I think that services like timer.h and cache.h are generic RTOS services where this makes sense. I'm not sure about interrupt.h which seems like a very driver specific. And alloc.h seems like a third category. There we probably want a SOF side layer on top of the RTOS.

Oh well, let's work from this.

File: src/audio/buffer.c (hunk 6 above)
Comment (lgirdwood) [reply src/audio/buffer.c:10]: agree, but I need to de-tangle before I can remove. This PR and others in progress are the de-tangling.

File: zephyr/include/rtos/alloc.h (hunk 7 above)
Comment (lgirdwood) [reply zephyr/include/rtos/alloc.h:17]: This is going towards a more POSIX like API, but it need some de-tangling before that can happen.

**Pull request merged**